### ISSUE TITLE
Go 1.13 support

### ISSUE THREAD

**Issue by hajimehoshi** [gopherjs/gopherjs#929]
The season will come soon :)

**Comment by eclipseo** [comment]
We are already testing Go 1.13 in Fedora, and the latest build for F31 failed because of this.
https://koji.fedoraproject.org/koji/taskinfo?taskID=36633628

**Comment by dmitshur** [comment]
I've created the [`go1.13`](https://github.com/gopherjs/gopherjs/commits/go1.13) branch with the first commit. The rest of the work still needs to be done by someone.

**Comment by dmitshur** [comment]
I've pushed [one commit](https://github.com/gopherjs/gopherjs/commit/c6ec0c360b352a369b45f5096449d47b503f9383) to `go1.13` branch. It fixes build errors in `sync` package.

Next up we'll need deal with `internal/reflectlite`. It's a new internal package added in Go 1.13:

https://tip.golang.org/pkg/internal/reflectlite/

> Package reflectlite implements lightweight version of reflect, not using any package except for "runtime" and "unsafe".

We'll need to find a way to implement it. I tried two approaches unsuccessfully so far:

1. replacing all the `internal/reflectlite` imports with `reflect`:
  - this doesn't work because of import cycles, `reflect` imports `errors` which imports `internal/reflectlite`
2. copying `reflect` package source code (overrides) and trimming them down to what's needed
  - this is hard because there's no clean subset of `reflect` that will implement all of `internal/reflectlite` API; it needs to be done properly and will be a substantial effort, which requires being familiar with how GopherJS implements and represents reflection (something I'm not closely familiar with)

Stopping here for now, just wanted to post an update about some progress made. Others are welcome to pick it up from here if you have the bandwidth.

**Comment by inliquid** [comment]
Any ETA on this?

**Comment by benma** [comment]
Bump.

**Comment by myitcv** [comment]
FWIW I have an 80% working 1.13 branch at https://github.com/myitcv/gopherjs/pull/50

Tests are currently the only known outstanding item. i.e. `gopherjs test` needs some reworking. Otherwise, using `gopherjs serve` appears to work.

**Comment by paralin** [comment]
@myitcv I did some tests with my app against wasm. Unfortunately as of now without significant optimization work, wasm will quickly run out of memory and consumes a ton of CPU

I know that this will improve with work to the Go compiler as well as new features / design aspects being ironed out in WebAssembly. But it also seems to be a bit /too/ alpha to use for even a moderately critical application, while the JavaScript approach seems like it will work well in a lot of environments for quite some time to come.

For that reason I think the Go to Js approach will be useful and viable for a long time. Thank you for your work towards go1.13, I will test it and let you know if I find any quirks.

**Comment by nevkontakte** [comment]
Hi @myitcv I'm interested in making 1.13 support happen sooner. Anything in particular I could help with? :)

**Comment by paralin** [comment]
@myitcv There's currently over 1200 lines of code hidden in your branch in the "tool.go" and similar files in the root of the gopherjs repository. That code is not accessible by my code importing the "gopherjs/build" pakages directly. Would it be possible to move that common code used to generate the arguments for NewSession() (including imports) to a package where I can access it externaly?

**Comment by myitcv** [comment]
@nevkontakte - if you're happy to jump in and start fixing things up on https://github.com/myitcv/gopherjs/pull/50 then please do, just give me a heads up on what you're planning to work on. Otherwise I will get to doing bits and pieces as and when I can. Verifying your applications work with that branch is probably one of the biggest helps, save the fact that `gopherjs test` does not work per above.

@paralin following on from our conversation in https://github.com/gopherjs/gopherjs/issues/855, my primary focus has been on a minimal working solution. I think we can look to "fix" the API of `gopherjs/build` afterwards.

**Comment by paralin** [comment]
@myitcv No worries, I was just posting a comment about it to keep it in the discussion. I can always modify the code myself in the meantime anyway. I appreciate the work & will keep testing the CLI version for now.

**Comment by nevkontakte** [comment]
Thanks @myitcv, I'll see what I can help with. I'll continue with my questions on your PR if you don't mind.

**Comment by AnikHasibul** [comment]
Hello, can you give us an ETA?

**Comment by goodxp** [comment]
I got an error when compiling go code to  js. My code structure is like:

//main.go to compile to js
package main
import "abc"
func main() {
	// do something with abc
	// js.Global.Set(...)
}

//abc.go
package abc
import "bytes"
func TestGopher() bool {
	buf := new(bytes.Buffer)
	if buf != nil {
		return true
	}
	return false
}

When compile, gopherjs output: 
../../../../../usr/local/go/src/sync/waitgroup.go:93:36: too many arguments

My Go version is 1.13.5 and gopherjs version is 1.13-3, running on MacOS 10.15.1.

I am new to GopherJS and there could be something I missed to encounter this problem, and I spent hours to track it down. Pls anyone help me out.

**Comment by hajimehoshi** [comment]
Go 1.13 is not supported yet.

**Comment by chrisprobst** [comment]
Any updates?

**Comment by benma** [comment]
:pray: @myitcv @dmitshur any progress? Anything we can help with? I am already starting to use docker containers with a downgraded Go just to use GopherJS.

**Comment by flimzy** [comment]
> I am already starting to use docker containers with a downgraded Go just to use GopherJS.

Just FYI, this should not be necessary any longer. You can now run GopherJS 1.12 with Go 1.13. This is what I've been doing.

**Comment by SysLabIT** [comment]
> You can now run GopherJS 1.12 with Go 1.13.
```

$ gopherjs serve
GopherJS 1.12-3 requires a Go 1.12.x distribution, but found version go1.13.7
$ gopherjs version
GopherJS 1.12-3
$ go version
go version go1.13.7 linux/amd64

```

**Comment by flimzy** [comment]
Fixed with #1015
